gems/aws-sdk-core/lib/aws-sdk-core/plugins/checksum_algorithm.rb

```diff
@@ -466,52 +469,93 @@ def response_header_to_verify(headers, validation_list)
-          partial_bytes = orig_body_size % CHUNK_SIZE
-          chunked_body_size = n_full_chunks * (CHUNK_SIZE + 8)
-          chunked_body_size += partial_bytes.to_s(16).size + partial_bytes + 4 unless  partial_bytes.zero?
+          n_full_chunks = orig_body_size / @base_chunk_size
+          partial_bytes = orig_body_size % @base_chunk_size
+
+          full_chunk_overhead = @base_chunk_size.to_s(HEX_BASE).size + CHUNK_OVERHEAD
+          chunked_body_size = n_full_chunks * (@base_chunk_size + full_chunk_overhead)
+          chunked_body_size += partial_bytes.to_s(HEX_BASE).size + partial_bytes + 4 unless partial_bytes.zero?
           trailer_size = ChecksumAlgorithm.trailer_length(@algorithm, @location_name)
           chunked_body_size + trailer_size
         end
 
         def rewind
           @io.rewind
+          @encoded_buffer = +''
+          @eof = false
+          @digest = ChecksumAlgorithm.digest_for_algorithm(@algorithm)
         end
 
-        def read(length, buf = nil)
-          # account for possible leftover bytes at the end, if we have trailer bytes, send them
-          if @trailer_io
-            return @trailer_io.read(length, buf)
-          end
+        def read(length = nil, buf = nil)
+          return '' if length&.zero?
+          return if eof?
 
-          chunk = @io.read(length)
-          if chunk
-            @digest.update(chunk)
-            application_chunked = "#{chunk.bytesize.to_s(16)}\r\n#{chunk}\r\n"
-            return StringIO.new(application_chunked).read(application_chunked.size, buf)
+          buf&.clear
+          output_buffer = buf || +''
+
+          fill_encoded_buffer(length)
+
+          if length
+            output_buffer << @encoded_buffer.slice!(0, length)
           else
-            trailers = {}
-            trailers[@location_name] = @digest.base64digest
-            trailers = trailers.map { |k,v| "#{k}:#{v}" }.join("\r\n")
-            @trailer_io = StringIO.new("0\r\n#{trailers}\r\n\r\n")
-            chunk = @trailer_io.read(length, buf)
+            output_buffer << @encoded_buffer
+            @encoded_buffer.clear
           end
-          chunk
+
+          output_buffer.empty? && eof? ? nil : output_buffer
+        end
+
+        def eof?
+          @eof && @encoded_buffer.empty?
+        end
+
+        private
+
+        def calculate_overhead(chunk_size)
+          chunk_size.to_s(HEX_BASE).size + CHUNK_OVERHEAD
+        end
+
+        def fill_encoded_buffer(required_length)
+          return if required_length && @encoded_buffer.bytesize >= required_length
+
+          while !@eof && fill_data?(required_length)
+            chunk = @io.read(@base_chunk_size)
+            if chunk && !chunk.empty?
+              @digest.update(chunk)
+              @encoded_buffer << "#{chunk.bytesize.to_s(HEX_BASE)}\r\n#{chunk}\r\n"
+            else
+              @encoded_buffer << "0\r\n#{trailer_string}\r\n\r\n"
+              @eof = true
+              break
```

Comment:
nit: since the while loop has a condition on @eof, you probably don't need the break here